src/types/pick.ts

```diff
@@ -0,0 +1,139 @@
+        const details: MessageDetails[] = [];
+        for (const [key, innerType] of this.propsArray) {
+            const missingKey = !hasOwnProperty(input, key);
+            const partialKey = this.propsInfo[key]?.partial;
+
+            if (missingKey) {
+                partialKey || details.push({ kind: 'missing property', property: key, type: innerType });
+                continue;
+            }
+
+            const innerResult = innerType.validate(input[key], options);
+            if (innerResult.ok) {
+                constructResult[key] = innerResult.value;
+            } else {
+                details.push(...prependPathToDetails(innerResult, key));
+            }
+        }
+        return this.createResult(input, options.mode === 'construct' ? constructResult : input, details);
+    }
+    override accept<R>(visitor: Visitor<R>): R {
+        return visitor.visitObjectLikeType(this);
+    }
+    /** {@inheritdoc BaseTypeImpl.maybeStringify} */
+    override maybeStringify(value: ResultType): string {
+        return interfaceStringify(this.propsArray, value as Record<string, unknown>);
+    }
+}
+define(PickType, 'basicType', 'object');
+
+export function pick<TypeImpl extends PickableImpl, Keys extends PickableKeys<TypeOf<TypeImpl>>>(
+    ...args: [name: string, baseType: TypeImpl, keys: OneOrMore<Keys>] | [baseType: TypeImpl, keys: OneOrMore<Keys>]
+): ObjectType<DistributedPick<TypeOf<TypeImpl>, Keys>> {
+    const [name, baseType, keys] = decodeOptionalName(args);
+    if (baseType instanceof UnionType) {
+        const unionType: UnionType<OneOrMore<BaseObjectLikeTypeImpl<TypeOf<TypeImpl>>>, TypeOf<TypeImpl>> = baseType;
+        if (!(unionType.basicType === 'object')) {
+            throw new Error(`Can only pick elements of unions with 'object' as basic type.`);
+        }
+        const narrowedKeys = narrowPickedKeys(keys, unionType.types);
+        if (!validPick(keys, narrowedKeys)) {
+            throw new Error('Selected keys describe impossible union variant.');
+        }
+
+        const innerTypes = unionType.types
+            .map((innerType, idx) => {
+                const specificKeys = narrowedKeys[idx] ?? [];
+                const keyCount = specificKeys.length;
+                // Only build the pick union elements if at least one of their keys has been picked.
+                // Also temporarily store the key count in a tuple to sort.
+                return [keyCount, keyCount ? new PickType(innerType, checkOneOrMore(specificKeys)) : undefined] as const;
+            })
+            // Sorting is necessary so the union type will internally match with a maximum subtype. This is so that types don't accidentally
+            // get narrowed too much during parsing and stringifying.
+            .sort(([aKeys, _1], [bKeys, _2]) => bKeys - aKeys) // Reverse sort: from most keys to least keys.
+            .flatMap(([_, innerType]) => innerType ?? []);
```

Comment:
Feels kinda nasty, but I didn't want to touch the implementation of `UnionType` to deal with finding the maximum pick type.

`UnionType` seems to just find the first matching type when parsing and stringifying, so when that happens to be a narrower type in the union, you get unexpected results (see the `Pick<UnionAC, 'a' | 'd'>` test case). The second conversion would also strip `d`, which is weird. Same would happen with stringify.

Reply:
Interesting idea. If we do something like this, it should be part of the `UnionType` implementation. I agree that this is a major footgun, so I'm currently considering starting with a `pick` operation on simpler interface types only.

I need to think and test your idea and possible other solutions to this footgun. We have had situations in our codebase where this happened, so I'm really happy if we can find a way to prevent "accidents" because of this.